* evpn primary address advertisement

chiragshah6 opened the pull request:
The objective of this PR is to announce type-5 routes and self type-2 routes in a EVPN symmetric routing MLAG deployment, with the switch’s individual (primary) IP as the next hop IP and the individual (system) MAC as router mac (RMAC) while leaving the behavior unchanged. 
The current behavior picks up the next hop IP from the VxLAN interface’s local tunnel IP and RMAC from L3-VNI SVI interface's MAC (anycast MAC).  This can lead to scenario where traffic forward to a switch which does not have a route and needs to make additional hop to go over peerlink. 

To achieve individual system MAC in conjunction with anycast MAC per VLAN/VNI basis, additional interface mac-vlan (VRR) interface required to be created. The SVI interface is configured with individual MAC and mac-vlan interface is configured with anycast MAC. 
To achieve individual system IP, the router-id of the bgp default instance is used.

User can enable/disable PIP feature, configure system IP and or configure system mac via cli
`[no] advertise-pip [<system-ip> [<system-mac>]]`
User configured value takes precedence and advertise along with type-5 and type-2 self routes.
By default, advertise-pip is enabled per bgp vrf instance and user can disable the feature via  cli  in that case existing anycast mac is used as RMAC and vxlan tunnel IP is used as nexthop. 

Testing:

type-5 route with system-IP 27.0.0.11 and system Mac 00:02:00:00:00:2e as RMAC when feature is enaled.
```
*> [5]:[0]:[24]:[81.6.1.0]
                    27.0.0.11                0             0 5541 i
                    ET:8 RT:5546:4002 Rmac:00:02:00:00:00:2e

TORC11# show bgp l2vpn evpn vni 4002
VNI: 4002 (known to the kernel)
  Type: L3
  Tenant VRF: vrf2
  RD: 141.2.1.2:7
  Originator IP: 36.0.0.11
  Advertise-gw-macip : n/a
  Advertise-pip: Yes
  System-IP: 27.0.0.11
  System-MAC: 00:02:00:00:00:2e
  Router-MAC: 44:38:39:ff:ff:01
  Import Route Target:
    5546:4002
  Export Route Target:
    5546:4002

```

srimohans commented:
@riw777 I will review this.

srimohans commented:
@chiragshah6 Are you planning for a quick explanation on this ?

chiragshah6 commented:
@srimohans Yes Mohan, I have reached out to you last week on multiple attempts. Today's meeting if time permits I will do prezo on it.

chiragshah6 commented:
last commit is rebase to latest master to resolve merge conflicts.

chiragshah6 commented on bgpd/bgp_evpn_private.h:
```diff
@@ -188,6 +188,16 @@ struct bgp_evpn_info {
+	/* PIP IP (sys ip) */
+	struct in_addr pip_ip;
+	struct in_addr pip_ip_static;
+	/* PIP MAC (sys MAC) */
```
For evpn routes there isnt' such case.

pguibert6WIND replied:
underlay ipv6 is not supported in current frr implementation ? can you confirm ?

chiragshah6 replied:
That is correct, the vtep-ip is supported for ipv4.  this case pip_ip is used as underlay nexthop and only supported format is ipv4.

riw777 replied:
IPv6 in the underlay should, eventually, be supported, but it currently isn't, AFAIK.

chiragshah6 commented on bgpd/bgp_evpn_vty.c:
```diff
@@ -5488,6 +5611,23 @@ void bgp_config_write_evpn_info(struct vty *vty, struct bgp *bgp, afi_t afi,
+			if (bgp->evpn_info->pip_ip_static.s_addr != INADDR_ANY)
+				vty_out(vty, "  advertise-pip ip %s",
+				inet_ntoa(bgp->evpn_info->pip_ip_static));
+			if (!is_zero_mac(&(bgp->evpn_info->pip_rmac_static))) {
```
convert to inet_ntop in latest patch

chiragshah6 commented on bgpd/bgp_evpn.c:
```diff
@@ -1543,11 +1543,44 @@ static int update_evpn_type5_route(struct bgp *bgp_vrf, struct prefix_evpn *evp,
+			   prefix2str(evp, buf1, sizeof(buf1)),
+			   prefix_mac2str(&attr.rmac, buf, sizeof(buf)),
+			   inet_ntoa(attr.nexthop));
+	}
```
convert to inet_ntop in latest patch

chiragshah6 commented on zebra/zebra_vxlan.c:
```diff
@@ -3735,6 +3737,52 @@ static struct interface *zvni_map_to_svi(vlanid_t vid, struct interface *br_if)
+
+	/* Identify corresponding VLAN interface. */
+	zns = zebra_ns_lookup(NS_DEFAULT);
+	for (rn = route_top(zns->if_table); rn; rn = route_next(rn)) {
```
Can be taken as separate item?

pguibert6WIND replied:
ok. this is just a reminder for both of us.

chiragshah6 commented on zebra/zebra_vxlan.c:
```diff
@@ -8527,6 +8527,67 @@ int zebra_vxlan_svi_up(struct interface *ifp, struct interface *link_if)
+
+	link_if = if_lookup_by_index_per_ns(zebra_ns_lookup(NS_DEFAULT),
+					    link_zif->link_ifindex);
+
```
Can be changed later with few other places require the conversion within zebra_vxlan.c

pguibert6WIND replied:
agree

pguibert6WIND commented on doc/user/bgp.rst:
```diff
@@ -1995,6 +1995,61 @@ address-family:
+disable the feature via configuration CLI. Once the feature is disable or
+MAC-VLAN interface is not configured, all routes follow the same behavior
+of using same next-hop and RMAC values.
+
```
adding a configuration example could have helped.

chiragshah6 commented:
Rebase to resolve merge conflicts. 

@pguibert6WIND  I have addressed all the comments and keeping the tab of the couple of open items relates to netns and will submit the PR once netns pr merges.

Can you please merge this PR if no major concerns left?
